<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>singleseqgset: Single-cell RNAseq gene set enrichment analysis • singleseqgset</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="singleseqgset: Single-cell RNAseq gene set enrichment analysis">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">singleseqgset</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.1.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/singleseqgset.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>singleseqgset: Single-cell RNAseq gene set enrichment analysis</h1>
            
      
      
      <div class="hidden name"><code>singleseqgset.Rmd</code></div>

    </div>

    
    
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>Gene set enrichment analysis is a ubiquitious tool used for extracting biologically meaningful information from a genomic dataset. There are many different flavors of tools available for gene set enrichment analysis, but the one most frequently encountered in the wild is the pioneering work of Subramanian et al, PNAS 2005.</p>
<p>They key concept in any gene set enrichment analysis is to compare a metric (e.g. log fold change in gene expression between groups) for genes within a set versus genes outside of a set. Gene sets can be obtained from many locations, including the Molecular Signatures Database (MSigDB) at the Broad and The Gene Ontology Resource. It is important to select gene sets that are best able to answer the biological question(s) at hand. For example, you generally would not use a gene set from the Broad’s C7 (Immunologic Gene Sets) to answer questions about the development of neurons (unless you have a really good reason!).</p>
<p>The approch to gene set enrichment that I have developed here is derived from the work of Di Wu and Gordon K Smyth, Nucleic Acids Research 2012. Wu and Smyth developed a Correlation Adjusted MEan RAnk gene set test (CAMERA). CAMERA is a competitive gene set enrichment test that controls for intergene correlation within the gene set. I have chosen to use CAMERA because it does not rely on the assumption of the independence of genes (since we know that genes often have structured co-expression patterns), nor does it rely on permutations of gene labels or resampling of test samples. CAMERA controls for inter-gene correlation by generating a variance inflation factor based on the degree of inter-gene correlation, and incorporates this variance inflation into a Wilcoxon rank sum test. I recommend checking out the CAMERA publication if you are into the math.</p>
<p>This vignette demonstrates a standard use of singleseqgset on simulated data. We will go through the following steps:</p>
<ul>
<li>Simulate expression data using the R package <em>splatter</em>
</li>
<li>Download gene sets of interest using <em>msigdbr</em>
</li>
<li>Add specific gene sets to our simulated data</li>
<li>Process our data using a standard Seurat workflow (v.2.3.4)</li>
<li>Use singleseqgset to perform gene set enrichment analysis</li>
<li>Plot the results in a heatmap</li>
</ul>
</div>
<div id="simulate-data-using-splatter" class="section level1">
<h1 class="hasAnchor">
<a href="#simulate-data-using-splatter" class="anchor"></a>Simulate data using splatter</h1>
<p>First, we load up all the necessary pacakges for this vignette, and simulate data to use with splatter.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/message">suppressMessages</a></span>({</a>
<a class="sourceLine" id="cb1-2" data-line-number="2"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(splatter)</a>
<a class="sourceLine" id="cb1-3" data-line-number="3"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(Seurat)</a>
<a class="sourceLine" id="cb1-4" data-line-number="4"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(msigdbr)</a>
<a class="sourceLine" id="cb1-5" data-line-number="5"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(singleseqgset)</a>
<a class="sourceLine" id="cb1-6" data-line-number="6"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(heatmap3)</a>
<a class="sourceLine" id="cb1-7" data-line-number="7">})</a>
<a class="sourceLine" id="cb1-8" data-line-number="8"></a>
<a class="sourceLine" id="cb1-9" data-line-number="9"><span class="co">#Create parameters and simulate data</span></a>
<a class="sourceLine" id="cb1-10" data-line-number="10">sc.params &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/splatter/topics/newParams">newSplatParams</a></span>(<span class="dt">nGenes=</span><span class="dv">1000</span>,<span class="dt">batchCells=</span><span class="dv">5000</span>)</a>
<a class="sourceLine" id="cb1-11" data-line-number="11">sim.groups &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/splatter/topics/splatSimulate">splatSimulate</a></span>(<span class="dt">params=</span>sc.params,<span class="dt">method=</span><span class="st">"groups"</span>,<span class="dt">group.prob=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="fl">0.4</span>,<span class="fl">0.2</span>,<span class="fl">0.3</span>,<span class="fl">0.1</span>),<span class="dt">de.prob=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="fl">0.20</span>,<span class="fl">0.20</span>,<span class="fl">0.1</span>,<span class="fl">0.3</span>),<span class="dt">verbose=</span>F)</a>
<a class="sourceLine" id="cb1-12" data-line-number="12">sim.groups <span class="co">#Check out the SingleCellExperiment object with simulated dataset</span></a></code></pre></div>
<pre><code>## class: SingleCellExperiment 
## dim: 1000 5000 
## metadata(1): Params
## assays(6): BatchCellMeans BaseCellMeans ... TrueCounts counts
## rownames(1000): Gene1 Gene2 ... Gene999 Gene1000
## rowData names(8): Gene BaseGeneMean ... DEFacGroup3 DEFacGroup4
## colnames(5000): Cell1 Cell2 ... Cell4999 Cell5000
## colData names(4): Cell Batch Group ExpLibSize
## reducedDimNames(0):
## spikeNames(0):</code></pre>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1"><span class="kw">colData</span>(sim.groups) <span class="co">#The colData column "Group" contains the groups of simulated cells</span></a></code></pre></div>
<pre><code>## DataFrame with 5000 rows and 4 columns
##              Cell       Batch    Group       ExpLibSize
##          &lt;factor&gt; &lt;character&gt; &lt;factor&gt;        &lt;numeric&gt;
## Cell1       Cell1      Batch1   Group1 73324.2901799195
## Cell2       Cell2      Batch1   Group1 71115.1438815874
## Cell3       Cell3      Batch1   Group3  89689.371004738
## Cell4       Cell4      Batch1   Group3  44896.460028516
## Cell5       Cell5      Batch1   Group3 53956.0668720417
## ...           ...         ...      ...              ...
## Cell4996 Cell4996      Batch1   Group2 79041.7558615305
## Cell4997 Cell4997      Batch1   Group4 66684.3797711752
## Cell4998 Cell4998      Batch1   Group3 70407.9613848431
## Cell4999 Cell4999      Batch1   Group3 64645.3257427214
## Cell5000 Cell5000      Batch1   Group3 62266.9119469426</code></pre>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1"><span class="co">#We will pull the simulated counts and groups from the sim.groups object</span></a>
<a class="sourceLine" id="cb5-2" data-line-number="2">sim.counts &lt;-<span class="st"> </span><span class="kw">assays</span>(sim.groups)<span class="op">$</span>counts</a>
<a class="sourceLine" id="cb5-3" data-line-number="3">groups &lt;-<span class="st"> </span><span class="kw">colData</span>(sim.groups)<span class="op">$</span>Group</a>
<a class="sourceLine" id="cb5-4" data-line-number="4"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(groups) &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">rownames</a></span>(<span class="kw">colData</span>(sim.groups))</a>
<a class="sourceLine" id="cb5-5" data-line-number="5"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/table">table</a></span>(groups)</a></code></pre></div>
<pre><code>## groups
## Group1 Group2 Group3 Group4 
##   1977    992   1536    495</code></pre>
<p>We have created some simulated data consisting of 4 clusters, with various fractions of genes differentially expressed between clusters.</p>
</div>
<div id="download-gene-sets-of-interest-using-msigdbr" class="section level1">
<h1 class="hasAnchor">
<a href="#download-gene-sets-of-interest-using-msigdbr" class="anchor"></a>Download gene sets of interest using msigdbr</h1>
<p>Here, we will use the misigdbr package to download the Human Hallmark Gene Sets. Alternatively, we could download gene sets directly from the Broad’s website: <a href="http://software.broadinstitute.org/gsea/msigdb/index.jsp" class="uri">http://software.broadinstitute.org/gsea/msigdb/index.jsp</a>, and read them into R using the <em>GSEABase</em> package.</p>
<p>Note: It is imperative to choose the proper gene annotation style (i.e. either gene symbols or entrez gene IDs; here we just gene symbols) and species (here we use human) for your project. Otherwise, the gene names in your gene sets will not match those in your data, and all gene sets will recieve a “0” enrichment score.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" data-line-number="1">h.human &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/msigdbr/topics/msigdbr">msigdbr</a></span>(<span class="dt">species=</span><span class="st">"Homo sapiens"</span>,<span class="dt">category=</span><span class="st">"H"</span>)</a>
<a class="sourceLine" id="cb7-2" data-line-number="2"></a>
<a class="sourceLine" id="cb7-3" data-line-number="3">h.names &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/unique">unique</a></span>(h.human<span class="op">$</span>gs_name)</a>
<a class="sourceLine" id="cb7-4" data-line-number="4"></a>
<a class="sourceLine" id="cb7-5" data-line-number="5">h.sets &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/vector">vector</a></span>(<span class="st">"list"</span>,<span class="dt">length=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/length">length</a></span>(h.names))</a>
<a class="sourceLine" id="cb7-6" data-line-number="6"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(h.sets) &lt;-<span class="st"> </span>h.names</a>
<a class="sourceLine" id="cb7-7" data-line-number="7"></a>
<a class="sourceLine" id="cb7-8" data-line-number="8"><span class="cf">for</span> (i <span class="cf">in</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(h.sets)) {</a>
<a class="sourceLine" id="cb7-9" data-line-number="9">    h.sets[[i]] &lt;-<span class="st"> </span><span class="kw">pull</span>(h.human[h.human<span class="op">$</span>gs_name<span class="op">==</span>i,<span class="st">"gene_symbol"</span>])</a>
<a class="sourceLine" id="cb7-10" data-line-number="10">}</a></code></pre></div>
</div>
<div id="add-gene-sets-to-simulated-clusters" class="section level1">
<h1 class="hasAnchor">
<a href="#add-gene-sets-to-simulated-clusters" class="anchor"></a>Add gene sets to simulated clusters</h1>
<p>We are going to assign 5 gene sets to be exclusively expressed in each cluster. We will select these 20 sets randomly, and each gene will be drawn from a zero-inflated Poisson distribution with the probability of success equal to 50% and a lambda value of 10. This will simulate moderate dropout and relatively low gene counts of expression. Please pardon the dense code for simulating gene sets!</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" data-line-number="1">sets.to.use &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/sample">sample</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/seq">seq</a></span>(<span class="dv">1</span>,<span class="dv">50</span>,<span class="dv">1</span>),<span class="dv">20</span>,<span class="dt">replace=</span>F)</a>
<a class="sourceLine" id="cb8-2" data-line-number="2">sets.and.groups &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/data.frame">data.frame</a></span>(<span class="dt">set=</span>sets.to.use,<span class="dt">group=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste</a></span>(<span class="st">"Group"</span>,<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/rep">rep</a></span>(<span class="dv">1</span><span class="op">:</span><span class="dv">4</span>,<span class="dt">each=</span><span class="dv">5</span>),<span class="dt">sep=</span><span class="st">""</span>))</a>
<a class="sourceLine" id="cb8-3" data-line-number="3"></a>
<a class="sourceLine" id="cb8-4" data-line-number="4"><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/nrow">nrow</a></span>(sets.and.groups)) {</a>
<a class="sourceLine" id="cb8-5" data-line-number="5"></a>
<a class="sourceLine" id="cb8-6" data-line-number="6">  set.to.use &lt;-<span class="st"> </span>sets.and.groups[i,<span class="st">"set"</span>]</a>
<a class="sourceLine" id="cb8-7" data-line-number="7">  group.to.use &lt;-<span class="st"> </span>sets.and.groups[i,<span class="st">"group"</span>]</a>
<a class="sourceLine" id="cb8-8" data-line-number="8">  gene.set.length &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/length">length</a></span>(h.sets[[set.to.use]])</a>
<a class="sourceLine" id="cb8-9" data-line-number="9">  blank.matrix &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/matrix">matrix</a></span>(<span class="dv">0</span>,<span class="dt">ncol=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/nrow">ncol</a></span>(sim.counts),<span class="dt">nrow=</span>gene.set.length)</a>
<a class="sourceLine" id="cb8-10" data-line-number="10">  <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">rownames</a></span>(blank.matrix) &lt;-<span class="st"> </span>h.sets[[sets.to.use[i]]]</a>
<a class="sourceLine" id="cb8-11" data-line-number="11">  matching &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">rownames</a></span>(blank.matrix) <span class="op">%in%</span><span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">rownames</a></span>(sim.counts)</a>
<a class="sourceLine" id="cb8-12" data-line-number="12"></a>
<a class="sourceLine" id="cb8-13" data-line-number="13">  <span class="cf">if</span> (<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/any">any</a></span>(matching<span class="op">==</span><span class="ot">TRUE</span>)) {</a>
<a class="sourceLine" id="cb8-14" data-line-number="14"></a>
<a class="sourceLine" id="cb8-15" data-line-number="15">    matching.genes &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">rownames</a></span>(blank.matrix)[matching]</a>
<a class="sourceLine" id="cb8-16" data-line-number="16">    nonmatching.genes &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/sets">setdiff</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">rownames</a></span>(blank.matrix),matching.genes)</a>
<a class="sourceLine" id="cb8-17" data-line-number="17">    blank.matrix &lt;-<span class="st"> </span>blank.matrix[<span class="op">!</span>matching,]</a>
<a class="sourceLine" id="cb8-18" data-line-number="18">    sim.counts &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/cbind">rbind</a></span>(sim.counts,blank.matrix)</a>
<a class="sourceLine" id="cb8-19" data-line-number="19"></a>
<a class="sourceLine" id="cb8-20" data-line-number="20">  } <span class="cf">else</span> {</a>
<a class="sourceLine" id="cb8-21" data-line-number="21"></a>
<a class="sourceLine" id="cb8-22" data-line-number="22">    sim.counts &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/cbind">rbind</a></span>(sim.counts,blank.matrix)</a>
<a class="sourceLine" id="cb8-23" data-line-number="23">    matching.genes &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">rownames</a></span>(blank.matrix)</a>
<a class="sourceLine" id="cb8-24" data-line-number="24"></a>
<a class="sourceLine" id="cb8-25" data-line-number="25">  }</a>
<a class="sourceLine" id="cb8-26" data-line-number="26"></a>
<a class="sourceLine" id="cb8-27" data-line-number="27">  group.cells &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">colnames</a></span>(sim.counts)[groups<span class="op">==</span>group.to.use]</a>
<a class="sourceLine" id="cb8-28" data-line-number="28"></a>
<a class="sourceLine" id="cb8-29" data-line-number="29">  <span class="cf">for</span> (z <span class="cf">in</span> group.cells) {</a>
<a class="sourceLine" id="cb8-30" data-line-number="30"></a>
<a class="sourceLine" id="cb8-31" data-line-number="31">    <span class="cf">if</span> (<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/any">any</a></span>(matching<span class="op">==</span><span class="ot">TRUE</span>)) {</a>
<a class="sourceLine" id="cb8-32" data-line-number="32"></a>
<a class="sourceLine" id="cb8-33" data-line-number="33">      sim.counts[matching.genes,z] &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/ifelse">ifelse</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/Binomial">rbinom</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/length">length</a></span>(matching.genes),<span class="dt">size=</span><span class="dv">1</span>,<span class="dt">prob=</span><span class="fl">0.5</span>)<span class="op">&gt;</span><span class="dv">0</span>,<span class="dv">0</span>,<span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/Poisson">rpois</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/length">length</a></span>(matching.genes),<span class="dt">lambda=</span><span class="dv">10</span>))</a>
<a class="sourceLine" id="cb8-34" data-line-number="34">      sim.counts[nonmatching.genes,z] &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/ifelse">ifelse</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/Binomial">rbinom</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/length">length</a></span>(nonmatching.genes),<span class="dt">size=</span><span class="dv">1</span>,<span class="dt">prob=</span><span class="fl">0.5</span>)<span class="op">&gt;</span><span class="dv">0</span>,<span class="dv">0</span>,<span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/Poisson">rpois</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/length">length</a></span>(nonmatching.genes),<span class="dt">lambda=</span><span class="dv">10</span>))</a>
<a class="sourceLine" id="cb8-35" data-line-number="35"></a>
<a class="sourceLine" id="cb8-36" data-line-number="36">    } <span class="cf">else</span> {</a>
<a class="sourceLine" id="cb8-37" data-line-number="37"></a>
<a class="sourceLine" id="cb8-38" data-line-number="38">    sim.counts[matching.genes,z] &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/ifelse">ifelse</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/Binomial">rbinom</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/length">length</a></span>(matching.genes),<span class="dt">size=</span><span class="dv">1</span>,<span class="dt">prob=</span><span class="fl">0.5</span>)<span class="op">&gt;</span><span class="dv">0</span>,<span class="dv">0</span>,<span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/Poisson">rpois</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/length">length</a></span>(matching.genes),<span class="dt">lambda=</span><span class="dv">10</span>))</a>
<a class="sourceLine" id="cb8-39" data-line-number="39"></a>
<a class="sourceLine" id="cb8-40" data-line-number="40">    }</a>
<a class="sourceLine" id="cb8-41" data-line-number="41">  }</a>
<a class="sourceLine" id="cb8-42" data-line-number="42"></a>
<a class="sourceLine" id="cb8-43" data-line-number="43">}</a>
<a class="sourceLine" id="cb8-44" data-line-number="44"></a>
<a class="sourceLine" id="cb8-45" data-line-number="45"><span class="co">#Here, we will check out the sum of expression for the first gene set</span></a>
<a class="sourceLine" id="cb8-46" data-line-number="46">len.set1 &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/length">length</a></span>(h.sets[[sets.to.use[[<span class="dv">1</span>]]]])</a>
<a class="sourceLine" id="cb8-47" data-line-number="47"><span class="kw"><a href="https://www.rdocumentation.org/packages/graphics/topics/plot">plot</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/apply">apply</a></span>(sim.counts[<span class="dv">1001</span><span class="op">:</span>(<span class="dv">1000</span><span class="op">+</span>len.set1),],<span class="dv">2</span>,sum),<span class="dt">xlim=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dv">0</span>,<span class="dv">5000</span>),<span class="dt">xlab=</span><span class="st">"Cells"</span>,<span class="dt">ylab=</span><span class="st">"Sum of Gene Set 1 Expression"</span>)</a></code></pre></div>
<p><img src="singleseqgset_files/figure-html/unnamed-chunk-3-1.png" width="672" style="display: block; margin: auto;"></p>
<p>We have successfully modified the simulated count matrix to contain gene sets of interest in specific clusters. For example, we know that Group1 cells should show gene set enrichment for:</p>
<p>HALLMARK_APICAL_SURFACE, HALLMARK_COMPLEMENT, HALLMARK_DNA_REPAIR, HALLMARK_IL2_STAT5_SIGNALING, HALLMARK_UNFOLDED_PROTEIN_RESPONSE</p>
</div>
<div id="seurat-workflow-on-simulated-data" class="section level1">
<h1 class="hasAnchor">
<a href="#seurat-workflow-on-simulated-data" class="anchor"></a>Seurat workflow on simulated data</h1>
<p>Now that we’ve simlulated some data, we will pass it through a standard Seurat workflow (using v2.3.4) to visualize our simulated data. We will ignore the gene sets we added to the count matrix, and will pretend that our first 1000 genes are highly variable genes.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" data-line-number="1"><span class="co">#Pass simulated data to Seurat</span></a>
<a class="sourceLine" id="cb9-2" data-line-number="2">ser &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/Seurat/topics/CreateSeuratObject">CreateSeuratObject</a></span>(<span class="dt">raw.data=</span>sim.counts)</a>
<a class="sourceLine" id="cb9-3" data-line-number="3"></a>
<a class="sourceLine" id="cb9-4" data-line-number="4"><span class="co">#Normal Seurat workflow</span></a>
<a class="sourceLine" id="cb9-5" data-line-number="5">ser &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/Seurat/topics/NormalizeData">NormalizeData</a></span>(ser)</a>
<a class="sourceLine" id="cb9-6" data-line-number="6">ser<span class="op">@</span>var.genes &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">rownames</a></span>(ser<span class="op">@</span>raw.data)[<span class="dv">1</span><span class="op">:</span><span class="dv">1000</span>]</a>
<a class="sourceLine" id="cb9-7" data-line-number="7">ser &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/Seurat/topics/ScaleData">ScaleData</a></span>(ser,<span class="dt">genes.use=</span>ser<span class="op">@</span>var.genes)</a></code></pre></div>
<pre><code>## Scaling data matrix</code></pre>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" data-line-number="1"><span class="co">#We will pretend like the 1000 simulated genes are the "variable genes",</span></a>
<a class="sourceLine" id="cb11-2" data-line-number="2"><span class="co">#and we will skip FindVariableGenes from Seurat</span></a>
<a class="sourceLine" id="cb11-3" data-line-number="3">ser &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/Seurat/topics/RunPCA">RunPCA</a></span>(ser,<span class="dt">pc.genes=</span>ser<span class="op">@</span>var.genes,<span class="dt">do.print=</span><span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb11-4" data-line-number="4"></a>
<a class="sourceLine" id="cb11-5" data-line-number="5"><span class="kw"><a href="https://www.rdocumentation.org/packages/Seurat/topics/PCElbowPlot">PCElbowPlot</a></span>(ser)</a></code></pre></div>
<p><img src="singleseqgset_files/figure-html/unnamed-chunk-4-1.png" width="672" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb12-1" data-line-number="1"><span class="co">#Looks like the first 4 PCs capture most of the variance</span></a>
<a class="sourceLine" id="cb12-2" data-line-number="2"><span class="co">#We will include the first 5</span></a>
<a class="sourceLine" id="cb12-3" data-line-number="3"></a>
<a class="sourceLine" id="cb12-4" data-line-number="4">ser &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/Seurat/topics/RunTSNE">RunTSNE</a></span>(ser,<span class="dt">dims.use=</span><span class="dv">1</span><span class="op">:</span><span class="dv">5</span>)</a>
<a class="sourceLine" id="cb12-5" data-line-number="5"></a>
<a class="sourceLine" id="cb12-6" data-line-number="6"><span class="co">#Add the simulated data IDs to the Seurat object</span></a>
<a class="sourceLine" id="cb12-7" data-line-number="7">data.to.add &lt;-<span class="st"> </span><span class="kw">colData</span>(sim.groups)<span class="op">$</span>Group</a>
<a class="sourceLine" id="cb12-8" data-line-number="8"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(data.to.add) &lt;-<span class="st"> </span>ser<span class="op">@</span>cell.names</a>
<a class="sourceLine" id="cb12-9" data-line-number="9">ser &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/Seurat/topics/AddMetaData">AddMetaData</a></span>(ser,<span class="dt">metadata=</span>data.to.add,<span class="dt">col.name=</span><span class="st">"real_id"</span>)</a>
<a class="sourceLine" id="cb12-10" data-line-number="10">ser &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/Seurat/topics/SetAllIdent">SetAllIdent</a></span>(ser,<span class="dt">id=</span><span class="st">"real_id"</span>)</a>
<a class="sourceLine" id="cb12-11" data-line-number="11"></a>
<a class="sourceLine" id="cb12-12" data-line-number="12"><span class="co">#We will skip clustering with Seurat because we already know the true clusters IDs</span></a>
<a class="sourceLine" id="cb12-13" data-line-number="13"></a>
<a class="sourceLine" id="cb12-14" data-line-number="14"><span class="kw"><a href="https://www.rdocumentation.org/packages/Seurat/topics/TSNEPlot">TSNEPlot</a></span>(ser,<span class="dt">group.by=</span><span class="st">"real_id"</span>)</a></code></pre></div>
<p><img src="singleseqgset_files/figure-html/unnamed-chunk-4-2.png" width="672" style="display: block; margin: auto;"></p>
</div>
<div id="use-singleseqgset-to-perform-gene-set-enrichment-analysis" class="section level1">
<h1 class="hasAnchor">
<a href="#use-singleseqgset-to-perform-gene-set-enrichment-analysis" class="anchor"></a>Use singleseqgset to perform gene set enrichment analysis</h1>
<p>We will now get to the point of the vignette: gene set enrichment analysis!</p>
<p>First, we will calculate our metric for the gene set enrichment test, which in this case is log fold change between clusters for all genes. We choose to use the log-normalized data that have been corrected for library size. This is stored in the “<span class="citation">@data</span>” slot of Seurat.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb13-1" data-line-number="1">logfc.data &lt;-<span class="st"> </span><span class="kw"><a href="../reference/logFC.html">logFC</a></span>(<span class="dt">cluster.ids=</span>ser<span class="op">@</span>meta.data<span class="op">$</span>real_id,<span class="dt">expr.mat=</span>ser<span class="op">@</span>data)</a>
<a class="sourceLine" id="cb13-2" data-line-number="2"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(logfc.data)</a></code></pre></div>
<pre><code>## [1] "cluster.cells"  "log.fc.cluster"</code></pre>
<p>The logFC function returns a list of length 2, containing the cells in each cluster and the log fold changes in genes between the cluster of interest and all other clusters. This data is required for the next step, where we calcuate enrichment scores and p values.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb15-1" data-line-number="1">gse.res &lt;-<span class="st"> </span><span class="kw"><a href="../reference/wmw_gsea.html">wmw_gsea</a></span>(<span class="dt">expr.mat=</span>ser<span class="op">@</span>data,<span class="dt">cluster.cells=</span>logfc.data[[<span class="dv">1</span>]],<span class="dt">log.fc.cluster=</span>logfc.data[[<span class="dv">2</span>]],<span class="dt">gene.sets=</span>h.sets)</a></code></pre></div>
<pre><code>## [1] "Removed 1 rows with all z-scores equal to zero."</code></pre>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb17-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(gse.res)</a></code></pre></div>
<pre><code>## [1] "GSEA_statistics" "GSEA_p_values"</code></pre>
<p>This function returns a two item list, the first containing the test statistics “GSEA_statistics” and the second containing the p values “GSEA_p_values”. We can plot these results as a heatmap the visualize differentially regulated gene sets.</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb19-1" data-line-number="1">res.stats &lt;-<span class="st"> </span>gse.res[[<span class="st">"GSEA_statistics"</span>]]</a>
<a class="sourceLine" id="cb19-2" data-line-number="2">res.pvals &lt;-<span class="st"> </span>gse.res[[<span class="st">"GSEA_p_values"</span>]]</a>
<a class="sourceLine" id="cb19-3" data-line-number="3"></a>
<a class="sourceLine" id="cb19-4" data-line-number="4">res.pvals &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/apply">apply</a></span>(res.pvals,<span class="dv">2</span>,p.adjust,<span class="dt">method=</span><span class="st">"fdr"</span>) <span class="co">#Correct for multiple comparisons</span></a>
<a class="sourceLine" id="cb19-5" data-line-number="5"></a>
<a class="sourceLine" id="cb19-6" data-line-number="6">res.stats[<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/order">order</a></span>(res.stats[,<span class="dv">1</span>],<span class="dt">decreasing=</span><span class="ot">TRUE</span>)[<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>],] <span class="co">#Top gene sets enriched by z scores</span></a></code></pre></div>
<pre><code>##                                              Group1      Group2    Group3
## HALLMARK_COMPLEMENT                       5.1574883 -0.09158513 -3.665589
## HALLMARK_IL2_STAT5_SIGNALING              5.0244223 -2.35815436 -3.294866
## HALLMARK_DNA_REPAIR                       4.7507909 -1.60640697 -2.170185
## HALLMARK_UNFOLDED_PROTEIN_RESPONSE        4.4252275 -3.43523859 -2.794865
## HALLMARK_APICAL_SURFACE                   3.3986861 -0.73663025  0.000000
## HALLMARK_BILE_ACID_METABOLISM             0.0000000 -4.76489163 -3.147855
## HALLMARK_NOTCH_SIGNALING                  0.0000000  0.00000000  0.000000
## HALLMARK_PEROXISOME                       0.0000000  0.00000000 -2.281640
## HALLMARK_REACTIVE_OXIGEN_SPECIES_PATHWAY  0.0000000  0.00000000 -1.624385
## HALLMARK_ANGIOGENESIS                    -0.6272414 -1.75751992  0.000000
##                                              Group4
## HALLMARK_COMPLEMENT                      -7.6289460
## HALLMARK_IL2_STAT5_SIGNALING             -3.6373362
## HALLMARK_DNA_REPAIR                      -3.4498250
## HALLMARK_UNFOLDED_PROTEIN_RESPONSE       -1.9363310
## HALLMARK_APICAL_SURFACE                   0.0000000
## HALLMARK_BILE_ACID_METABOLISM            -0.1628354
## HALLMARK_NOTCH_SIGNALING                  0.5230535
## HALLMARK_PEROXISOME                      -0.3289919
## HALLMARK_REACTIVE_OXIGEN_SPECIES_PATHWAY -0.6633797
## HALLMARK_ANGIOGENESIS                    -1.4628583</code></pre>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb21-1" data-line-number="1">res.pvals[<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/order">order</a></span>(res.stats[,<span class="dv">1</span>],<span class="dt">decreasing=</span><span class="ot">TRUE</span>)[<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>],] <span class="co">#Top gene sets by p values</span></a></code></pre></div>
<pre><code>##                                                Group1       Group2
## HALLMARK_COMPLEMENT                      1.362660e-06 1.000000e+00
## HALLMARK_IL2_STAT5_SIGNALING             2.474256e-06 3.333099e-02
## HALLMARK_DNA_REPAIR                      7.637311e-06 1.656575e-01
## HALLMARK_UNFOLDED_PROTEIN_RESPONSE       3.147129e-05 1.381409e-03
## HALLMARK_APICAL_SURFACE                  1.508095e-03 5.861109e-01
## HALLMARK_BILE_ACID_METABOLISM            1.000000e+00 7.715670e-06
## HALLMARK_NOTCH_SIGNALING                 1.000000e+00 1.000000e+00
## HALLMARK_PEROXISOME                      1.000000e+00 1.000000e+00
## HALLMARK_REACTIVE_OXIGEN_SPECIES_PATHWAY 1.000000e+00 1.000000e+00
## HALLMARK_ANGIOGENESIS                    5.776566e-01 1.287544e-01
##                                               Group3       Group4
## HALLMARK_COMPLEMENT                      0.000711277 3.865785e-13
## HALLMARK_IL2_STAT5_SIGNALING             0.002193161 6.427686e-04
## HALLMARK_DNA_REPAIR                      0.044534864 1.195067e-03
## HALLMARK_UNFOLDED_PROTEIN_RESPONSE       0.009422769 8.350102e-02
## HALLMARK_APICAL_SURFACE                  1.000000000 1.000000e+00
## HALLMARK_BILE_ACID_METABOLISM            0.003349134 9.274294e-01
## HALLMARK_NOTCH_SIGNALING                 1.000000000 7.181930e-01
## HALLMARK_PEROXISOME                      0.035581310 8.658554e-01
## HALLMARK_REACTIVE_OXIGEN_SPECIES_PATHWAY 0.127759613 6.211821e-01
## HALLMARK_ANGIOGENESIS                    1.000000000 2.130849e-01</code></pre>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb23-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(h.sets)[sets.to.use[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]] <span class="co">#Compare to the simulate sets we created</span></a></code></pre></div>
<pre><code>## [1] "HALLMARK_APICAL_SURFACE"           
## [2] "HALLMARK_COMPLEMENT"               
## [3] "HALLMARK_DNA_REPAIR"               
## [4] "HALLMARK_IL2_STAT5_SIGNALING"      
## [5] "HALLMARK_UNFOLDED_PROTEIN_RESPONSE"</code></pre>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb25-1" data-line-number="1"><span class="co">#Plot the z scores with heatmap3</span></a>
<a class="sourceLine" id="cb25-2" data-line-number="2"><span class="kw"><a href="https://www.rdocumentation.org/packages/heatmap3/topics/heatmap3">heatmap3</a></span>(res.stats,<span class="dt">Colv=</span><span class="ot">NA</span>,<span class="dt">cexRow=</span><span class="fl">0.5</span>,<span class="dt">cexCol=</span><span class="dv">1</span>,<span class="dt">scale=</span><span class="st">"row"</span>)</a></code></pre></div>
<p><img src="singleseqgset_files/figure-html/unnamed-chunk-7-1.png" width="960" style="display: block; margin: auto;"></p>
<p>Congratulations, you have performed your first gene set enrichment analysis with singleseqgset!</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#introduction">Introduction</a></li>
      <li><a href="#simulate-data-using-splatter">Simulate data using splatter</a></li>
      <li><a href="#download-gene-sets-of-interest-using-msigdbr">Download gene sets of interest using msigdbr</a></li>
      <li><a href="#add-gene-sets-to-simulated-clusters">Add gene sets to simulated clusters</a></li>
      <li><a href="#seurat-workflow-on-simulated-data">Seurat workflow on simulated data</a></li>
      <li><a href="#use-singleseqgset-to-perform-gene-set-enrichment-analysis">Use singleseqgset to perform gene set enrichment analysis</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Anthony Cillo.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
